packages/core/src/ui/UICanvas.ts

```diff
@@ -0,0 +1,495 @@
+        this._sortDistance = this._distance;
+      case CanvasRenderMode.WorldSpace:
+        this._sortDistance = Vector3.distance(cameraPosition, (this._transform as UITransform).worldPosition);
+    }
+  }
+
+  /**
+   * @internal
+   */
+  override _onEnableInScene(): void {
+    this._entity._dispatchModify(EntityModifyFlags.UICanvasEnableInScene);
+    this._entity._registerModifyListener(this._onEntityModify);
+    this._registerParentListener();
+    this._setIsRootCanvas(this._checkIsRootCanvas());
+  }
+
+  /**
+   * @internal
+   */
+  override _onDisableInScene(): void {
+    this._removeParentListener();
+    this._entity._removeModifyListener(this._onEntityModify);
+    this._entity._dispatchModify(EntityModifyFlags.UICanvasDisableInScene);
+    this._setIsRootCanvas(false);
+    this._renderers.length = 0;
+  }
+
+  /**
+   * @internal
+   */
+  rayCast(ray: Ray, out: HitResult, distance: number = Number.MAX_SAFE_INTEGER): boolean {
+    const { renderers } = this;
+    for (let i = renderers.length - 1; i >= 0; i--) {
+      const renderer = renderers[i];
+      if (renderer.rayCastAble && renderer._raycast(ray, out, distance)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private _adapterPoseInScreenSpace(): void {
+    const { _renderCamera: renderCamera, _transform: transform } = this;
+    if (renderCamera) {
+      const { transform: cameraTransform } = renderCamera.entity;
+      const { worldPosition: cameraWorldPosition, worldForward: cameraWorldForward } = cameraTransform;
+      const { _distance: distance } = this;
+      transform.setWorldPosition(
+        cameraWorldPosition.x + cameraWorldForward.x * distance,
+        cameraWorldPosition.y + cameraWorldForward.y * distance,
+        cameraWorldPosition.z + cameraWorldForward.z * distance
+      );
+      transform.worldRotationQuaternion.copyFrom(cameraTransform.worldRotationQuaternion);
+    } else {
+      const { canvas } = this.engine;
+      transform.setWorldPosition(canvas.width / 2, canvas.height / 2, 0);
+      transform.worldRotationQuaternion.set(0, 0, 0, 1);
+    }
+  }
+
+  private _adapterSizeInScreenSpace(): void {
+    const { _renderCamera: renderCamera } = this;
+    const { x: width, y: height } = this._referenceResolution;
+    let curWidth: number;
+    let curHeight: number;
+    if (renderCamera) {
+      curHeight = renderCamera.isOrthographic
+        ? renderCamera.orthographicSize * 2
+        : 2 * (Math.tan(MathUtil.degreeToRadian(renderCamera.fieldOfView / 2)) * this._distance);
+      curWidth = renderCamera.aspectRatio * curHeight;
+    } else {
+      const canvas = this.engine.canvas;
+      curHeight = canvas.height;
+      curWidth = canvas.width;
+    }
+    let expectX: number, expectY: number, expectZ: number;
+    switch (this._resolutionAdaptationStrategy) {
+      case ResolutionAdaptationStrategy.WidthAdaptation:
+        expectX = expectY = expectZ = curWidth / width;
```

Comment:
We can simplify the code by extracting the common calculations for curWidth / width and curHeight / height, and reduce the repetitive assignments for expectX, expectY, and expectZ. Here's a suggestion:

const aspectRatioX = curWidth / width;
const aspectRatioY = curHeight / height;

Reply:
There is no multiple calculation here, and using variables to save it will increase the number of temporary variables.